keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,612 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `v1.Endpoints` API has been essentially (though not _actually_)
+deprecated since EndpointSlices became GA in 1.21. Several new Service
+features (such as dual-stack and topology, not to mention "services
+with more than 1000 endpoints") are implemented only for
+EndpointSlice, not for Endpoints. Kube-proxy no longer uses Endpoints
+ever, for anything, and the Gateway API conformance tests also require
+implementations to use EndpointSlices, so Gateway implementations
+don't use Endpoints either.
+
+Despite this, kube-controller-manager still does all of the work of
+managing Endpoints objects for all Services, and a cluster cannot pass
+the conformance test suite unless the Endpoints and EndpointSlice
+Mirroring controllers are running, even though in many cases nothing
+will ever look at the output of the Endpoints controller (and the
+EndpointSlice Mirroring controller will never output anything).
+
+While Kubernetes's API guarantees make it essentially impossible to
+ever actually fully remove Endpoints, we should at least move toward a
+world where most users run Kubernetes with the Endpoints and
+EndpointSlice Mirroring controllers disabled.
+
+## Motivation
+
+### Goals
+
+- Officially declare v1.Endpoints to be deprecated. Update
+  documentation and put out appropriate communications (blog posts,
+  etc) to ensure that end users are aware of this deprecation.
+
+- Add warnings (e.g., `Warning:` headers on Endpoints create/update)
+  to alert users of the fact that Endpoints is deprecated.
+
+- Ensure that all core Kubernetes code uses EndpointSlices rather than
+  Endpoints.
+
+- Update the e2e test suite to make it possible to run it in a "no
+  Endpoints controller" configuration, by rewriting some tests and
+  adding feature tags to others.
+
+- Update the conformance test suite to not require the Endpoints and
+  EndpointSlice Mirroring controllers to be running, by rewriting some
+  tests and demoting others from conformance.
+
+- Explicitly document that disabling `endpoints-controller` and/or
+  `endpointslice-mirroring-controller` via kube-controller-manager's
+  `--controllers` flag is a supported and conforming configuration.
+
+- Come up with a plan to clean up stale Endpoints in clusters that
```

Comment:
Should this KEP not be this plan? Or are you saying that for a future stage of this KEP you will update this to have a plan?

Reply:
Yes, it's an UNRESOLVED point. We need a plan before the KEP is complete.

Reply:
Im seeing approvals from others. https://github.com/kubernetes/enhancements/pull/4975#issuecomment-2646659917

I should’ve said I was reviewing from PRR shadow viewpoint. Just curious if you think you’ll have this approved for enhancement freeze?